stan/math/rev/core/vari.hpp

```diff
@@ -210,9 +199,134 @@ class vari_view<T, require_all_t<bool_constant<!is_plain_type<T>::value>,
+   */
+  inline auto head(Eigen::Index n) {
+    return vari_view<decltype(val_.head(n))>(val_.head(n), adj_.head(n));
+  }
+
+  /**
+   * View of the tail of the Eigen vector types.
+   * @param n Number of elements to return from bottom of vector.
+   */
+  inline auto tail(Eigen::Index n) {
+    return vari_view<decltype(val_.tail(n))>(val_.tail(n), adj_.tail(n));
+  }
+
+  /**
+   * View block of N elements starting at position `i`
+   * @param i Starting position of block.
+   * @param n Number of elements in block
+   */
+  inline auto segment(Eigen::Index i, Eigen::Index n) {
+    return vari_view<decltype(val_.segment(i, n))>(val_.segment(i, n),
+                                                   adj_.segment(i, n));
+  }
+
+  /**
+   * View row of eigen matrices.
+   * @param i Row index to slice.
+   */
+  inline auto row(Eigen::Index i) {
+    return vari_view<decltype(val_.row(i))>(val_.row(i), adj_.row(i));
+  }
+
+  /**
+   * View column of eigen matrices
+   * @param i Column index to slice
+   */
+  inline auto col(Eigen::Index i) {
+    return vari_view<decltype(val_.col(i))>(val_.col(i), adj_.col(i));
+  }
+
+  /**
+   * Get coefficient of eigen matrices
+   * @param i Row index
+   * @param j Column index
+   */
+  inline auto coeff(Eigen::Index i, Eigen::Index j) {
+    return vari_value<double&>(val_.coeffRef(i, j), adj_.coeffRef(i, j));
+  }
+
+  /**
+   * Get coefficient of eigen matrices
+   * @param i Column index to slice
+   */
+  inline auto coeff(Eigen::Index i) {
+    return vari_value<double&>(val_.coeffRef(i), adj_.coeffRef(i));
+  }
+
+  /**
+   * Get coefficient of eigen matrices
+   * @param i Column index to slice
+   */
+  inline auto operator()(Eigen::Index i) { return this->coeff(i); }
+
+  /**
+   * Get coefficient of eigen matrices
+   * @param i Row index
+   * @param j Column index
+   */
+  inline auto operator()(Eigen::Index i, Eigen::Index j) {
+    return this->coeff(i, j);
+  }
+
+  /**
+   * Return an expression that operates on the rows of the matrix `vari`
+   */
+  inline auto rowwise() {
+    return vari_view<decltype(val_.rowwise())>(val_.rowwise(), adj_.rowwise());
+  }
+
+  /**
+   * Return an expression that operates on the columns of the matrix `vari`
+   */
+  inline auto colwise() {
+    return vari_view<decltype(val_.colwise())>(val_.colwise(), adj_.colwise());
+  }
+
+  /**
+   * Return an expression an expression to reverse the order of the coefficients
+   * inside of a `vari` matrix
+   */
+  inline auto reverse() {
+    return vari_view<decltype(val_.reverse())>(val_.reverse(), adj_.reverse());
+  }
+
+  /**
+   * Return the number of rows for this class's `val_` member
+   */
+  Eigen::Index rows() const { return val_.rows(); }
+  /**
+   * Return the number of columns for this class's `val_` member
+   */
+  Eigen::Index cols() const { return val_.cols(); }
+  /**
+   * Return the size of this class's `val_` member
+   */
+  Eigen::Index size() const { return val_.size(); }
```

Comment:
I think we could use a common base class for vari_view and Eigen vari_value to avoid duplicating this code.

Reply:
Yes! I wrote up a neat lil' crtp to do this I'll post up now